File: MyDailyBeat/API/VerveAPI/HealthDatabase.m
//
//  HealthDatabase.m
//  VerveAPI
//
//  Created by Virinchi Balabhadrapatruni on 5/28/16.
//  Copyright © 2016 eVerveCorp. All rights reserved.
//

#import "HealthDatabase.h"

@interface HealthDatabase()

@property (nonatomic, strong) NSString *docDir, *dbFilename;

@end

@implementation HealthDatabase

static HealthDatabase *_database;

+ (HealthDatabase*)database {
    if (_database == nil) {
        _database = [[HealthDatabase alloc] init];
    }
    return _database;
}

- (id) init {
    self = [super init];
    if (self) {
        NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
        self.docDir = [paths objectAtIndex:0];
        self.dbFilename = @"healthlist.sql";
        [self copyDatabaseIntoDocDir];
    }
    return self;
}

- (void) copyDatabaseIntoDocDir {
    NSString *dest = [self.docDir stringByAppendingPathComponent:self.dbFilename];
    if (![[NSFileManager defaultManager] fileExistsAtPath:dest]) {
        NSString *source = [[[NSBundle mainBundle] resourcePath] stringByAppendingPathComponent:self.dbFilename];
        NSError *error;
        [[NSFileManager defaultManager] copyItemAtPath:source toPath:dest error:&error];
        
        if (error != nil ) {
            NSLog(@"%@", [error localizedDescription]);
        }
    }
}

- (NSArray *)prescripProviders {
    NSMutableArray *retval = [[NSMutableArray alloc] init];
    NSString *sqLiteDb = [self.docDir stringByAppendingPathComponent:self.dbFilename];
    
    if (sqlite3_open([sqLiteDb UTF8String], &_database) != SQLITE_OK) {
        NSLog(@"Failed to open database!");
    } else {
        NSString *query = @"SELECT rowid, URL, logoURL FROM prescripProviders ORDER BY rowid ASC";
        sqlite3_stmt *statement;
        if (sqlite3_prepare_v2(_database, [query UTF8String], -1, &statement, nil)
            == SQLITE_OK) {
            while (sqlite3_step(statement) == SQLITE_ROW) {
                int uniqueId = sqlite3_column_int(statement, 0);
                char *nameChars = (char *) sqlite3_column_text(statement, 1);
                char *appChars = (char *) sqlite3_column_text(statement, 2);
                NSString *name = [[NSString alloc] initWithUTF8String:nameChars];
                NSString *appURL = [[NSString alloc] initWithUTF8String:appChars];
                PrescripProviderInfo *info = [[PrescripProviderInfo alloc] initWithUniqueId:uniqueId URL:name logoURL:appURL];
                [retval addObject:info];
            }
            sqlite3_finalize(statement);
        }
    }
    
    sqlite3_close(_database);
    
    return retval;
    
}

- (NSArray *)healthPortals {
    NSMutableArray *retval = [[NSMutableArray alloc] init];
    NSString *sqLiteDb = [self.docDir stringByAppendingPathComponent:self.dbFilename];
    
    if (sqlite3_open([sqLiteDb UTF8String], &_database) != SQLITE_OK) {
        NSLog(@"Failed to open database!");
    } else {
        NSString *query = @"SELECT rowid, URL, logoURL FROM healthPortals ORDER BY rowid ASC";
        sqlite3_stmt *statement;
        if (sqlite3_prepare_v2(_database, [query UTF8String], -1, &statement, nil)
            == SQLITE_OK) {
            while (sqlite3_step(statement) == SQLITE_ROW) {
                int uniqueId = sqlite3_column_int(statement, 0);
                char *urlChars = (char *) sqlite3_column_text(statement, 1);
                char *logoChars = (char *) sqlite3_column_text(statement, 2);
                NSString *URL = [[NSString alloc] initWithUTF8String:urlChars];
                NSString *logoURL = [[NSString alloc] initWithUTF8String:logoChars];
                HealthInfo *info = [[HealthInfo alloc] initWithUniqueId:uniqueId URL:URL logoURL:logoURL];
                [retval addObject:info];
            }
            sqlite3_finalize(statement);
        }
    }
    
    sqlite3_close(_database);
    
    return retval;
    
}

- (void) insertIntoDatabase: (HealthInfo *) healthPortalOrNil orPrescriptionProvider: (PrescripProviderInfo *) prescripProviderOrNil{
    NSString *sqLiteDb = [self.docDir stringByAppendingPathComponent:self.dbFilename];
    
    if (sqlite3_open([sqLiteDb UTF8String], &_database) != SQLITE_OK) {
        NSLog(@"Failed to open database!");
    } else {
        if (prescripProviderOrNil != nil) {
            // prescrip providers
            NSString *query = @"create table if not exists prescripProviders(URL text, logoURL text)";
            const char * qstring = [query UTF8String];
            char * errMsg;
            int prepareStatementResult = sqlite3_exec(_database, qstring, NULL, NULL, &errMsg);
            if (prepareStatementResult == SQLITE_OK) {
                NSString *query2 = [NSString stringWithFormat:@"insert into prescripProviders values('%@', '%@')", prescripProviderOrNil.URL, prescripProviderOrNil.logoURL];
                
                const char* qstring2 = [query2 UTF8String];
                NSLog(@"Query: %s", qstring2);
                char * errMsg2;
                
                // Load all data from database to memory.
                int prepareStatementResult2 = sqlite3_exec(_database, qstring2, NULL, NULL, &errMsg2);
                if(prepareStatementResult2 == SQLITE_OK) {
                    
                } else {
                    NSLog(@"Error: %s", errMsg2);
                }
            } else {
                NSLog(@"Error: %s", errMsg);
            }
        } else {
            // health portals
            NSString *query = @"create table if not exists healthPortals(URL text, logoURL text)";
            const char * qstring = [query UTF8String];
            char * errMsg;
            int prepareStatementResult = sqlite3_exec(_database, qstring, NULL, NULL, &errMsg);
            if (prepareStatementResult == SQLITE_OK) {
                NSString *query2 = [NSString stringWithFormat:@"insert into healthPortals values('%@', '%@')", healthPortalOrNil.URL, healthPortalOrNil.logoURL];
                
                const char* qstring2 = [query2 UTF8String];
                NSLog(@"Query: %s", qstring2);
                char * errMsg2;
                
                // Load all data from database to memory.
                int prepareStatementResult2 = sqlite3_exec(_database, qstring2, NULL, NULL, &errMsg2);
                if(prepareStatementResult2 == SQLITE_OK) {
                    
                } else {
                    NSLog(@"Error: %s", errMsg2);
                }
            } else {
                NSLog(@"Error: %s", errMsg);
            }
        }
        
        
    }
    
    sqlite3_close(_database);
    
}


@end

File: MDB_Swift/MyDailyBeat/API/src/HealthDatabase.swift
//  Converted with Swiftify v1.0.6242 - https://objectivec2swift.com/
//
//  HealthDatabase.swift
//  VerveAPI
//
//  Created by Virinchi Balabhadrapatruni on 5/28/16.
//  Copyright © 2016 eVerveCorp. All rights reserved.
//
import Foundation
public class HealthDatabase: NSObject {

    var docDir: String = ""
    var dbFilename: String = ""

    public func prescripProviders() -> [PrescripProviderInfo] {
        var retval = [PrescripProviderInfo]()
//        var sqLiteDb: String = URL(fileURLWithPath: self.docDir).appendingPathComponent(self.dbFilename).absoluteString
//        if sqlite3_open(sqLiteDb.utf8, self.database) != SQLITE_OK {
//            print("Failed to open database!")
//        }
//        else {
//            var query: String = "SELECT rowid, URL, logoURL FROM prescripProviders ORDER BY rowid ASC"
//            var statement: sqlite3_stmt?
//            if sqlite3_prepare_v2(self.database, query.utf8, -1, statement, nil) == SQLITE_OK {
//                while sqlite3_step(statement) == SQLITE_ROW {
//                    var uniqueId: Int = sqlite3_column_int(statement, 0)
//                    var nameChars = CChar(sqlite3_column_text(statement, 1))
//                    var appChars = CChar(sqlite3_column_text(statement, 2))
//                    var name = String(utf8String: nameChars)
//                    var appURL = String(utf8String: appChars)
//                    var info = PrescripProviderInfo(uniqueId: uniqueId, url: name, logoURL: appURL)
//                    retval.append(info)
//                }
//                sqlite3_finalize(statement)
//            }
//        }
//        sqlite3_close(self.database)
        return retval
    }

    public func healthPortals() -> [HealthInfo] {
        var retval = [HealthInfo]()
//        var sqLiteDb: String = URL(fileURLWithPath: self.docDir).appendingPathComponent(self.dbFilename).absoluteString
//        if sqlite3_open(sqLiteDb.utf8, self.database) != SQLITE_OK {
//            print("Failed to open database!")
//        }
//        else {
//            var query: String = "SELECT rowid, URL, logoURL FROM healthPortals ORDER BY rowid ASC"
//            var statement: sqlite3_stmt?
//            if sqlite3_prepare_v2(self.database, query.utf8, -1, statement, nil) == SQLITE_OK {
//                while sqlite3_step(statement) == SQLITE_ROW {
//                    var uniqueId: Int = sqlite3_column_int(statement, 0)
//                    var urlChars = CChar(sqlite3_column_text(statement, 1))
//                    var logoChars = CChar(sqlite3_column_text(statement, 2))
//                    var URL = String(utf8String: urlChars)
//                    var logoURL = String(utf8String: logoChars)
//                    var info = HealthInfo(uniqueId: uniqueId, url: URL, logoURL: logoURL)
//                    retval.append(info)
//                }
//                sqlite3_finalize(statement)
//            }
//        }
//        sqlite3_close(self.database)
        return retval
    }

    public func insert(intoDatabase healthPortalOrNil: HealthInfo?, orPrescriptionProvider prescripProviderOrNil: PrescripProviderInfo?) {
        var sqLiteDb: String = URL(fileURLWithPath: self.docDir).appendingPathComponent(self.dbFilename).absoluteString
//        if sqlite3_open(sqLiteDb.utf8, self.database) != SQLITE_OK {
//            print("Failed to open database!")
//        }
//        else {
//            if prescripProviderOrNil != nil {
//                    // prescrip providers
//                var query: String = "create table if not exists prescripProviders(URL text, logoURL text)"
//                let qstring = query.utf8
//                var errMsg: [CChar]
//                var prepareStatementResult: Int = sqlite3_exec(self.database, qstring, nil, nil, errMsg)
//                if prepareStatementResult == SQLITE_OK {
//                    var query2: String = "insert into prescripProviders values('\(prescripProviderOrNil.url)', '\(prescripProviderOrNil.logoURL)')"
//                    let qstring2 = query2.utf8
//                    var errMsg2: [CChar]
//                        // Load all data from database to memory.
//                    var prepareStatementResult2: Int = sqlite3_exec(self.database, qstring2, nil, nil, errMsg2)
//                    if prepareStatementResult2 == SQLITE_OK {
//
//                    }
//                    else {
//                        print("Error: \(errMsg2)")
//                    }
//                }
//                else {
//                    print("Error: \(errMsg)")
//                }
//            }
//            else {
//                    // health portals
//                var query: String = "create table if not exists healthPortals(URL text, logoURL text)"
//                let qstring = query.utf8
//                var errMsg: [CChar]
//                var prepareStatementResult: Int = sqlite3_exec(self.database, qstring, nil, nil, errMsg)
//                if prepareStatementResult == SQLITE_OK {
//                    var query2: String = "insert into healthPortals values('\(healthPortalOrNil.url)', '\(healthPortalOrNil.logoURL)')"
//                    let qstring2 = query2.utf8
//                    var errMsg2: [CChar]
//                        // Load all data from database to memory.
//                    var prepareStatementResult2: Int = sqlite3_exec(self.database, qstring2, nil, nil, errMsg2)
//                    if prepareStatementResult2 == SQLITE_OK {
//
//                    }
//                    else {
//                        print("Error: \(errMsg2)")
//                    }
//                }
//                else {
//                    print("Error: \(errMsg)")
//                }
//            }
//        }
//        sqlite3_close(self.database)
    }

    public static var database: HealthDatabase!

    override init() {
        super.init()
        
        var paths: [Any] = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)
        self.docDir = paths[0] as! String
        self.dbFilename = "healthlist.sql"
        self.copyIntoDocDir()
    
    }

    func copyIntoDocDir() {
        var dest: String = URL(fileURLWithPath: self.docDir).appendingPathComponent(self.dbFilename).absoluteString
        if !FileManager.default.fileExists(atPath: dest) {
            var source: String = URL(fileURLWithPath: (Bundle.main.resourcePath)!).appendingPathComponent(self.dbFilename).absoluteString
            var error: Error?
            try? FileManager.default.copyItem(atPath: source, toPath: dest)
            if error != nil {
                print("\(error?.localizedDescription)")
            }
        }
    }
}
